File: app/src/main/java/com/binatestation/kickstart/adapters/holders/BaseViewHolder.java
package com.binatestation.kickstart.adapters.holders;

import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.binatestation.kickstart.listeners.AdapterListener;
import com.binatestation.kickstart.listeners.ViewBinder;


/**
 * Created by RKR on 08-12-2017.
 * BaseViewHolder
 */

public abstract class BaseViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, ViewBinder {
    private final AdapterListener mAdapterListener;

    BaseViewHolder(View itemView, AdapterListener adapterListener) {
        super(itemView);
        this.mAdapterListener = adapterListener;
        itemView.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        int position = getAdapterPosition();
        if (position != RecyclerView.NO_POSITION && mAdapterListener != null && mAdapterListener.getClickListener() != null) {
            mAdapterListener.getClickListener().onClickItem(mAdapterListener.getItem(position), position, view);
        }
    }

    @Override
    public abstract void bindView(Object object);

    public AdapterListener getAdapterListener() {
        return mAdapterListener;
    }
}

File: kickoff/src/main/java/com/binatestation/android/kickoff/utils/adapters/holders/BaseViewHolder.kt
/*
 * (c) Binate Station Private Limited. All rights reserved.
 */

package com.binatestation.android.kickoff.utils.adapters.holders

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.binatestation.android.kickoff.utils.listeners.ViewBinder


/**
 * Created by RKR on 08-12-2017.
 * BaseViewHolder
 */

abstract class BaseViewHolder constructor(
    itemView: View
) : RecyclerView.ViewHolder(itemView), ViewBinder {

    private var onClickItem: ((position: Int, actionView: View) -> Unit)? = null

    fun setOnItemClickListener(onClickItem: (position: Int, actionView: View) -> Unit) {
        this.onClickItem = onClickItem
    }

    init {
        itemView.setOnClickListener { onClickItem?.let { onClick -> onClick(bindingAdapterPosition, it) } }
    }

    @Suppress("unused")
    fun setOnClick(view: View) {
        onClickItem?.let { onClick -> onClick(bindingAdapterPosition, view) }
    }

    abstract override fun bindView(`object`: Any?)

}
